Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator

zasdfgbnm opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33998 Kill CUDA_tensor_apply4
* **#33997 Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator**

Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

zasdfgbnm committed:
Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator

[ghstack-poisoned]

VitalyFedyunin reviewed (commented):
Please add benchmarks as this solution might impact performance.

zasdfgbnm commented:
@VitalyFedyunin Unfortunately there is a 30% perf drop as shown below. But TensorIterator does not support having two outputs, and it does not sound trivial to add this support, so we have to choose between:
1. Accepting the perf drop.
2. Being unable to remove `CUDA_tensor_apply4`.

If we choose 2, I think we at least need to move the `CUDA_tensor_apply4` to `Activations.cu` and rename it to `prelu_apply4` to discourage using it in other operators.

```python
import torch
print(torch.__version__)

# warm up
for i in range(1000):
    torch.arange(100000, device='cuda')
    
m = torch.nn.PReLU().cuda()
def benchmark(size):
    a = torch.randn(2 ** size, device='cuda')
    b = torch.randn_like(a)
    loss = m(a)
    torch.cuda.synchronize()
    %timeit loss.backward(b, retain_graph=True); torch.cuda.synchronize()
    
for i in range(15, 28):
    benchmark(i)
```

Before
```
1.5.0a0+9956a23
111 µs ± 1.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
123 µs ± 349 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
131 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
134 µs ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
160 µs ± 246 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
256 µs ± 503 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
459 µs ± 603 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
839 µs ± 418 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.6 ms ± 5.46 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.13 ms ± 5.23 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
6.22 ms ± 5.46 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
12.2 ms ± 5.26 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
24.3 ms ± 52.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

After
```
1.5.0a0+22d1de1
113 µs ± 943 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
121 µs ± 420 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
129 µs ± 686 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
142 µs ± 415 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
211 µs ± 433 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
345 µs ± 1.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
608 µs ± 548 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.13 ms ± 449 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.18 ms ± 3.86 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
4.37 ms ± 8.31 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
8.57 ms ± 4.22 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
17 ms ± 5.72 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
33.8 ms ± 50.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

zasdfgbnm committed:
Update on "Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator"


Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator"


Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator"


Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

[ghstack-poisoned]

VitalyFedyunin commented:
30% is too significant drop. let's stick with second proposal.

zasdfgbnm commented:
I will work on writing a kernel for prelu backward after this gets merged: https://github.com/pytorch/pytorch/pull/34749

zasdfgbnm committed:
Update on "Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator"


Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator"


Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

[ghstack-poisoned]

zasdfgbnm commented:
@VitalyFedyunin This is ready now, I write a kernel for prelu backward following the design of unrolled kernel in GPU loops. The benchmark is https://github.com/zasdfgbnm/things/blob/master/2020Q1/prelu-benchmark.ipynb (On V100)

On V100, the performance is slightly better, on RTX2080Ti, it is slightly worse.

VitalyFedyunin commented on aten/src/ATen/native/cuda/Loops.cuh:
```diff
@@ -9,6 +9,12 @@ namespace at { namespace native {
 constexpr int num_threads = C10_WARP_SIZE * 2;
 constexpr int thread_work_size = 4;
 constexpr int block_work_size = thread_work_size * num_threads;
+// Below is a macro version of the same thing, this is needed because
```
Please put them above and do something like: `constexpr int num_threads = NUM_THREADS;` to ensure we never have them defined differently, thanks.

zasdfgbnm replied:
fixed

zasdfgbnm committed:
Update on "Migrate prelu_cuda_backward from CUDA_tensor_apply4 to TensorIterator"


Differential Revision: [D20196786](https://our.internmc.facebook.com/intern/diff/D20196786)

[ghstack-poisoned]
